grails-app/services/org/pih/warehouse/core/IdentifierService.groovy

```diff
@@ -150,13 +153,83 @@ class IdentifierService {
         return identifier.toUpperCase()
     }
 
+    def generateSequentialProductIdentifier(ProductType productType, String sequentialPatternChar, Integer allowedSequences) {
```

Comment:
Could this be broken down into separate methods or does this need to be done in one unit of work?

validateSequenceNumberFormat() for the top part and generateSequenceNumber() in the bottom. This could probably be used by purchase order number and other sequence numbers.